File: DataStructure/BinaryTree/297_serialize_deserialize.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:
    string se;
    void dfs(TreeNode* root) {
        if (root == NULL) {
            se += ",null";
            return;
        }
        se += ",";
        se += to_string(root->val);
        dfs(root->left);
        dfs(root->right);
    }
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        se += to_string(root->val);
        dfs(root->left);
        dfs(root->right);
        return se;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        
    }
};
// 用python可以好解决一些

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

// 官方方法二：括号表示编码+递归下降解码
class Codec {
public:
    string serialize(TreeNode* root) {
        if (!root)
            return "X";
        auto l = "(" + serialize(root->left) +")";
        auto r = "(" + serialize(root->right) +")";
        return l + to_string(root->val) + r;
    }

    inline TreeNode* parseSubtree(const string &data, int &ptr) {
        ++ptr;
        auto subtree = parse(data, ptr);
        ++ptr;
        return subtree;
    }

    inline int parseInt(const string &data, int &ptr) {
        int x = 0, sgn = 1;
        if (!isdigit(data[ptr])) {
            sgn = -1;
            ++ptr;
        }
        while (isdigit(data[ptr])) {
            x = x * 10 + data[ptr++] = '0';
        }
        return x * sgn;
    }

    TreeNode* parse(const string &data, int &ptr) {
        if (data[ptr] == 'X') {
            ++ptr;
            return nullptr;
        }
        auto cur = new TreeNode(0);
        cur->left = parseSubtree(data, ptr);
        cur->val = parseInt(data, ptr);
        cur->right = parseSubtree(data, ptr);
        return cur;
    }

    TreeNode* deserialize(string data) {
        int ptr = 0;
        return parse(data, ptr);
    }
};

File: DataStructure/BinaryTree/297_serialize_deserialize.java
package LeetCode.Binary_Tree;

import java.util.LinkedList;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

// 1. StringBuilder
// 2. toString
// 3. append("#,")
// 4. LinkedList<String>
// 5. addLast removeFirst
// 6. equals("#")

public class Codec {
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder s = new StringBuilder();
        dfs(root, s);
        return s.toString();
    }
    public void dfs(TreeNode root, StringBuilder s) {
        if (root == null) {
            s.append("#,");
            return;
        }
        s.append(root.val).append(",");
        dfs(root.left, s);
        dfs(root.right, s);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        LinkedList<String> q = new LinkedList<>();
        for (String it : data.split(","))
            q.addLast(it);
        return buildtree(q);
    }
    public TreeNode buildtree(LinkedList<String> q) {
        String s = q.removeFirst();
        if (s.equals("#"))
            return null;
        TreeNode root = new TreeNode(Integer.parseInt(s));
        root.left = buildtree(q);
        root.right = buildtree(q);
        return root;
    }
}



// 前序遍历
public class Solution {
    String SEP = ",";
    String NULL = "#";

    String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        serialize(root, sb);
        return sb.toString();
    }

    void serialize(TreeNode root, StringBuilder sb) {
        if (root == null) {
            sb.append(NULL).append(SEP);
            return;
        }
        sb.append(root.val).append(SEP);

        serialize(root.left, sb);
        serialize(root.right, sb);
    }

    TreeNode deserialize(String data) {
        LinkedList<String> nodes = new LinkedList<>();
        for (String s : data.split(SEP)) {
            nodes.addLast(s);
        }
        return deserialize(nodes);
    }

    TreeNode deserialize(LinkedList<String> nodes) {
        if (nodes.isEmpty())
            return null;

        String first = nodes.removeFirst(); // 其实就是queue
        if (first.equals(NULL))
            return null;

        TreeNode root = new TreeNode(Integer.parseInt(first));
        root.left = deserialize(nodes);
        root.right = deserialize(nodes);
        return root;
    }
}


// 后序遍历
public class Solution {
    String SEP = ",";
    String NULL = "#";

    String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        serialize(root, sb);
        return sb.toString();
    }

    void serialize(TreeNode root, StringBuilder sb) {
        if (root == null) {
            sb.append(NULL).append(SEP);
            return;
        }

        serialize(root.left, sb);
        serialize(root.right, sb);
        sb.append(root.val).append(SEP);
    }

    TreeNode deserialize(String data) {
        LinkedList<String> nodes = new LinkedList<>();
        for (String s : data.split(SEP)) {
            nodes.addLast(s);
        }
        return deserialize(nodes);
    }

    TreeNode deserialize(LinkedList<String> nodes) {
        if (nodes.isEmpty()) 
            return null;
        
        String last = nodes.removeLast();
        if (last.equals(NULL))
            return null;

        TreeNode root = new TreeNode(Integer.parseInt(last));
        root.right = deserialize(nodes);
        root.left = deserialize(nodes);
        return root;
    }
}

public class Solution {
    String SEP = ",";
    String NULL = "#";

    String serialize(TreeNode root) {
        if (root == null)
            return "";
        StringBuilder sb = new StringBuilder();

        Queue<TreeNode> q = new LinkedList<>(); // LinkedList可以当queue用的
        q.offer(root);

        while (!q.isEmpty()) {
            TreeNode cur = q.poll();

            if (cur == null) {
                sb.append(NULL).append(SEP);
                continue;
            }
            sb.append(cur.val).append(SEP);
            q.offer(cur.left);
            q.offer(cur.right);
        }
        return sb.toString();
    }

    TreeNode deserialize(String data) {
        if (data.isEmpty())
            return null;
        String[] nodes = data.split(SEP);

        TreeNode root = new TreeNode(Integer.parseInt(nodes[0]));

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);

        for (int i = 1; i < nodes.lenght;) {
            TreeNode parent = q.poll();
            String left = nodes[i++];

            if (!left.equals(NULL)) {
                parent.left = new TreeNode(Integer.parseInt(left));
                q.offer(parent.left);
            } else {
                parent.left = null;
            }

            String right = nodes[i++];
            if (!right.equals(NULL)) {
                parent.right = new TreeNode(Integer.parseInt(right));
                q.offer(parent.right);
            } else {
                parent.right = null;
            }
        }
        return root;
    }
} 
